libraries/chain/block_header_state.cpp

```diff
@@ -89,6 +89,30 @@ finalizer_policy block_header_state::calculate_finalizer_policy(const finalizer_
    return result;
 }
 
+proposer_policy_diff block_header_state::calculate_proposer_policy_diff(const proposer_policy& new_policy) const {
+   if (proposer_policies.empty()) {
+      assert(active_proposer_policy);
+      return active_proposer_policy->create_diff(new_policy);
+   }
+   auto it = proposer_policies.rbegin();
+   assert(it != proposer_policies.rend());
+   return it->second->create_diff(new_policy);
+}
+
+proposer_policy block_header_state::calculate_proposer_policy(const proposer_policy_diff& diff) const {
```

Comment:
I think it the code would be nicer with a `get_last_proposed_proposerer_policy()`.

Indeed, the two calculate functions are barely needed anymore 

Same applies for finalizer_policy:

replace:

`new_finalizer_policy_diff = calculate_finalizer_policy_diff(*input.new_finalizer_policy);`

with 

`new_finalizer_policy_diff = get_last_proposed_finalizer_policy().create_diff(*input.new_finalizer_policy);`

and replace:

`finalizer_policy new_finalizer_policy = prev.calculate_finalizer_policy(*if_ext.new_finalizer_policy_diff);`

with 

`finalizer_policy new_finalizer_policy = prev.get_last_proposed_finalizer_policy().apply_diff(*if_ext.new_finalizer_policy_diff);`

(would need to update `apply_diff` to return a value).